MVP Procedural Scheduling UI

skovati opened the pull request:
```___REQUIRES_AERIE_PR___="1496"```

https://github.com/NASA-AMMOS/aerie/issues/1496

Adds a few small UI features in the scheduling goals panel that add additional functionality (arguments, duplicating goal invocations) when dealing with procedures instead of EDSL goals.

As part of the procedural scheduling backend changes, the primary key for `scheduling_specification_goals` is a new (auto-generated) `goal_invocation_id` int. This PR also updates a few functions / GQL queries to use this ID instead of `(spec, goal_id, revision)` when possible

mattdailis committed:
Prototype for showing parameters for scheduling procedures

mattdailis committed:
Add support for scheduling goal invocations

skovati committed:
wip goal invocation PK change

skovati committed:
fix revision bug

skovati committed:
fix scheduling goal page bug for non-procedural goals

duranb commented on src/components/scheduling/SchedulingGoalsPanel.svelte:
```diff
@@ -40,7 +40,7 @@
 
   // TODO: remove this after db merge as it becomes redundant
-  $: visibleSchedulingGoalSpecs = $allowedSchedulingGoalSpecs.filter(({ goal_metadata: goalMetadata }) => {
+  $: visibleSchedulingGoalSpecs = $schedulingGoalInvocations.filter(({ goal_metadata: goalMetadata }) => {
```
Does `schedulingGoalInvocations` only contain goals that the user has access to similar to `allowedSchedulingGoalSpecs`?

skovati replied:
Change has been made, `allowedSchedulingGoalInvocations` now exists and follows the same filtering pattern.

duranb commented on src/components/scheduling/goals/SchedulingGoal.svelte:
```diff
@@ -42,6 +51,36 @@
 
+  $: {
+    let revision = null;
+    console.log({ goalPlanSpec });
```
leftover console.log

skovati replied:
Done

duranb commented on src/components/scheduling/goals/SchedulingGoal.svelte:
```diff
@@ -42,6 +51,36 @@
+      Object.entries(schema.items).forEach(([name, subschema], i) => {
+        result.push({
+          errors: null,
+          // file?: File,
```
are these comments needed?

skovati replied:
Nope! Deleted them

duranb commented on src/components/scheduling/goals/SchedulingGoal.svelte:
```diff
@@ -87,12 +126,34 @@
   }
 
+  function duplicateGoalInvocation() {
+    dispatch('duplicateGoalInvocation', {
```
the `dispatch` type declaration should be updated to include these new events

skovati replied:
Change has been made

duranb reviewed (changes requested):
There are a few typing and linting changes that are necessary. @skovati let me know if VSCode isn't exposing those to you somehow.

duranb commented on src/types/metadata.ts:
```diff
@@ -12,6 +12,7 @@ export type BaseDefinition = {
   plans_using: Pick<Plan, 'id'>[];
   revision: number;
   tags: { tag: Tag }[];
+  type: "JAR" | "EDSL";
```
does this `type` apply to all other definition types (i.e. constraints, scheduling conditions) as well?

skovati replied:
The goal is eventually to have procedural constraints and conditions, but currently we just support procedural scheduling goals. I've made the change to just add this field to `SchedulingGoalDefinition` for now.

duranb commented on src/utilities/gql.ts:
```diff
@@ -2745,6 +2749,34 @@ const gql = {
 
+  SUB_SCHEDULING_GOAL_INVOCATIONS: `#graphql
+    subscription SubSchedulingGoalInvocations($planId: Int!) {
+      scheduling_specification_goals (where: {specification: {plan_id: {_eq: $planId}}}) {
```
Is `scheduling_specification_goals` the correct query? If so please reuse the `Queries.SCHEDULING_SPECIFICATION_GOALS` enum.

skovati replied:
Yep! Change has been made!

skovati committed:
fix linting errors

skovati committed:
move `type` field from `BaseDefinition` to `SchedulingGoalDefinition`

skovati committed:
Clean up SchedulingGoal.svelte with small suggestions

skovati committed:
run prettier

skovati committed:
oh eslint you are so picky

skovati committed:
fix svelte check errors

skovati committed:
run prettier

skovati committed:
follow pattern of filtering allowed goal invocations

skovati committed:
use `Queries.SCHEDULING_SPECIFICATION_GOALS` enum

skovati committed:
run prettier

skovati committed:
fix duplicate invocation bug when exiting the goal management modal

two part fix:
 - `allowedSchedulingGoalSpecs` was really an array, but was treated as an object
   - use `find` instead of indexing with `goal_id`
 - remove owner check, since that was always being triggered as true,
   causing an update
   - this extranous update was fine before, since the upsert would
     handle the deduplication, but now with goal invocations, where we
     allow duplicates, it was causing random duplications of goals when
     exiting the modal

skovati committed:
make eslint and prettier happy

duranb commented on src/components/scheduling/SchedulingGoalsPanel.svelte:
```diff
@@ -80,22 +80,54 @@
+    } = event;
+
+    if ($plan) {
+      await effects.updateSchedulingGoalPlanSpecifications(
```
Would `updateSchedulingGoalPlanSpecification` be more appropriate to use here since only one is being updated?

duranb replied:
Actually maybe this needs to be it's own separate `effect` function. If we call either, the error in the toast won't reflect the action taken. This isn't necessarily updating a spec, but more creating a new one. There's currently a function commented out in the effects file called `createSchedulingGoalPlanSpecification` that you might be able to repurpose.

duranb commented on src/components/scheduling/SchedulingGoalsPanel.svelte:
```diff
@@ -80,22 +80,54 @@
+      detail: { goal_metadata, specification_id, ...goalPlanSpec },
+    } = event;
+
+    console.log({ goalPlanSpec });
```
leftover console

duranb commented on src/utilities/gql.ts:
```diff
@@ -3436,7 +3487,7 @@ const gql = {
 
   UPDATE_SCHEDULING_GOAL_PLAN_SPECIFICATIONS: `#graphql
-    mutation UpdateSchedulingGoalPlanSpecifications($goalSpecsToUpdate: [scheduling_specification_goals_insert_input!]!, $goalSpecIdsToDelete: [Int!]! = [], $specificationId: Int!) {
+    mutation UpdateSchedulingGoalPlanSpecifications($goalSpecsToUpdate: [scheduling_specification_goals_insert_input!]!, $goalSpecIdsToDelete: [Int!]! = []) {
```
Because the pkey was changed, this query is no longer "correct". Can we change `updateSchedulingGoalPlanSpecifications` to `insertSchedulingGoalPlanSpecifications`, and remove the `on_conflict` specification? The effect function using this query will also need to update its reference to `updateSchedulingGoalPlanSpecifications`

duranb commented on src/components/scheduling/goals/SchedulingGoal.svelte:
```diff
@@ -42,6 +53,32 @@
+    let schema = version?.parameter_schema;
+    let result: any[] = [];
+    if (schema && schema.type === 'struct') {
+      Object.entries(schema.items).forEach(([name, subschema], i) => {
```
Rather than pushing into a `result` array typed as `any`, can this be structured like
```
if (schema && schema.type === 'struct') {
   formParameters = Object.entries(schema.items).map([name, subschema], i) => { 
     ...
    return { ... };
   });
} else {
  formParameters = [];
}
```

duranb commented on src/stores/scheduling.ts:
```diff
@@ -156,6 +162,10 @@ export const allowedSchedulingGoalSpecs: Readable<SchedulingGoalPlanSpecificatio
     $schedulingGoalSpecifications.filter(({ goal_metadata: goalMetadata }) => goalMetadata !== null),
 );
 
+export const allowedSchedulingGoalInvocations = derived([schedulingGoalInvocations], ([$schedulingGoalInvocations]) =>
```
I think this might be redundant. This doesn't seem different from `allowedSchedulingGoalSpecs` right?

skovati commented:
Visual bug seen in demo: new activities badge isn't keyed per invocation id, instead it applies that badge to all invocations sharing a goal id

skovati committed:
remove console.log

skovati committed:
remove unnecessary mutable result variable

skovati committed:
remove redundant invocation stores

skovati committed:
add type to plan spec mega query

skovati committed:
add new effect for creating goal invocation

skovati committed:
remove on_conflict from `updateSchedulingGoalPlanSpecifications`

skovati committed:
run prettier

duranb commented on src/components/scheduling/SchedulingGoalsPanel.svelte:
```diff
@@ -80,22 +80,48 @@
+      await effects.createSchedulingGoalPlanSpecification(
+        {
+          ...goalPlanSpec,
+          goal_invocation_id: undefined, // let DB autogenerate
```
My only concern with setting `goal_invocation_id` to undefined here is that there's a chance that we may forget to set it to `undefined` in other places that use `effects.createSchedulingGoalPlanSpecification`. It might be more appropriate to explicitly make the `SchedulingGoalPlanSpecInsertInput` type not take a `goal_invocation_id`.

skovati replied:
Good point. I've made a new `SchedulingGoalPlanSpecInvocationInsertInput` that `Omit`'s `goal_invocation_id`, since `SchedulingGoalPlanSpecInsertInput` gets used by other effects that need to update specs, and would therefore need to specific a specific invocation id

skovati committed:
add new invocation insert type that omits `goal_invocation_id`

skovati committed:
follow insert/set naming convention now that update doesn't use the input type

duranb reviewed (approved):
lgtm! thank you!!

skovati commented:
We'll want to merge https://github.com/NASA-AMMOS/aerie/pull/1496 first FYI

duranb commented:
Sounds good! Feel free to merge whenever you're ready.

dandelany commented on src/components/modals/ManagePlanSchedulingGoalsModal.svelte:
```diff
@@ -217,10 +217,13 @@
+
+          // if we find at least one goal invocation with the selected goal_id, we don't want to insert this goal_id into the plan spec
+          // i.e. this goal was already selected when we entered the modal, so we don't want to kick off an update, which would cause a duplicate invocation to appear
+          const goalAlreadyExistsInPlanSpec = $allowedSchedulingGoalSpecs.find(e => e.goal_id === goalId) !== undefined;
```
For V2 it would be nice to have a better UI for adding a second invocation of the same goal from the "Manage" modal... Currently since it's a checkbox, you can only add/remove a single invocation from here & can only add a 2nd via right-click -> "Duplicate Goal"

dandelany reviewed (approved):
👍 I've been testing this with the `feature/scheduling-upload` branch and looks ready for first release, thanks
